File: examples/android/app/src/main/java/com/example/noisemobile/NoiseSession.kt
package com.example.noisemobile

import java.nio.ByteBuffer

/**
 * Noise session mode
 */
enum class NoiseMode(val value: Int) {
    INITIATOR(0),
    RESPONDER(1)
}

/**
 * Noise protocol exceptions
 */
sealed class NoiseException(message: String) : Exception(message) {
    class InvalidParameter : NoiseException("Invalid parameter provided")
    class OutOfMemory : NoiseException("Out of memory")
    class HandshakeFailed : NoiseException("Handshake failed")
    class EncryptionFailed : NoiseException("Encryption failed")
    class DecryptionFailed : NoiseException("Decryption failed")
    class BufferTooSmall(val needed: Int) : NoiseException("Buffer too small, need $needed bytes")
    class InvalidState : NoiseException("Operation invalid in current state")
    class ProtocolError : NoiseException("Protocol error")
    class Unknown(val code: Int) : NoiseException("Unknown error: $code")
}

/**
 * Kotlin wrapper for Noise protocol sessions
 */
class NoiseSession private constructor(
    private val nativeHandle: Long
) : AutoCloseable {
    
    companion object {
        init {
            System.loadLibrary("noise_mobile")
        }
        
        /**
         * Create a new Noise session
         */
        @JvmStatic
        fun create(mode: NoiseMode): NoiseSession {
            val handle = when (mode) {
                NoiseMode.INITIATOR -> nativeCreateInitiator()
                NoiseMode.RESPONDER -> nativeCreateResponder()
            }
            
            if (handle == 0L) {
                throw NoiseException.OutOfMemory()
            }
            
            return NoiseSession(handle)
        }
        
        /**
         * Create a new Noise session with a specific private key
         */
        @JvmStatic
        fun createWithKey(mode: NoiseMode, privateKey: ByteArray): NoiseSession {
            if (privateKey.size != 32) {
                throw NoiseException.InvalidParameter()
            }
            
            val handle = when (mode) {
                NoiseMode.INITIATOR -> nativeCreateInitiatorWithKey(privateKey)
                NoiseMode.RESPONDER -> nativeCreateResponderWithKey(privateKey)
            }
            
            if (handle == 0L) {
                throw NoiseException.OutOfMemory()
            }
            
            return NoiseSession(handle)
        }
        
        // Native method declarations
        @JvmStatic
        private external fun nativeCreateInitiator(): Long
        
        @JvmStatic
        private external fun nativeCreateResponder(): Long
        
        @JvmStatic
        private external fun nativeCreateInitiatorWithKey(privateKey: ByteArray): Long
        
        @JvmStatic
        private external fun nativeCreateResponderWithKey(privateKey: ByteArray): Long
    }
    
    private var closed = false
    
    /**
     * Get the local public key
     */
    val publicKey: ByteArray
        get() {
            checkNotClosed()
            return nativeGetPublicKey(nativeHandle)
                ?: throw NoiseException.InvalidState()
        }
    
    /**
     * Check if handshake is complete
     */
    val isHandshakeComplete: Boolean
        get() {
            checkNotClosed()
            return nativeIsHandshakeComplete(nativeHandle)
        }
    
    /**
     * Write a handshake or transport message
     */
    fun writeMessage(payload: ByteArray = byteArrayOf()): ByteArray {
        checkNotClosed()
        
        val result = nativeWriteMessage(nativeHandle, payload)
        return result ?: throw getLastError()
    }
    
    /**
     * Read a handshake or transport message
     */
    fun readMessage(message: ByteArray): ByteArray {
        checkNotClosed()
        
        val result = nativeReadMessage(nativeHandle, message)
        return result ?: throw getLastError()
    }
    
    /**
     * Encrypt a message (transport mode only)
     */
    fun encrypt(plaintext: ByteArray): ByteArray {
        checkNotClosed()
        
        val result = nativeEncrypt(nativeHandle, plaintext)
        return result ?: throw getLastError()
    }
    
    /**
     * Decrypt a message (transport mode only)
     */
    fun decrypt(ciphertext: ByteArray): ByteArray {
        checkNotClosed()
        
        val result = nativeDecrypt(nativeHandle, ciphertext)
        return result ?: throw getLastError()
    }
    
    /**
     * Close the session and free resources
     */
    override fun close() {
        if (!closed && nativeHandle != 0L) {
            nativeDestroy(nativeHandle)
            closed = true
        }
    }
    
    private fun checkNotClosed() {
        if (closed) {
            throw IllegalStateException("Session is closed")
        }
    }
    
    private fun getLastError(): NoiseException {
        val errorCode = nativeGetLastError(nativeHandle)
        return when (errorCode) {
            1 -> NoiseException.InvalidParameter()
            2 -> NoiseException.OutOfMemory()
            3 -> NoiseException.HandshakeFailed()
            4 -> NoiseException.EncryptionFailed()
            5 -> NoiseException.DecryptionFailed()
            6 -> NoiseException.BufferTooSmall(0) // TODO: Get actual needed size
            7 -> NoiseException.InvalidState()
            8 -> NoiseException.ProtocolError()
            else -> NoiseException.Unknown(errorCode)
        }
    }
    
    // Native methods
    private external fun nativeDestroy(handle: Long)
    private external fun nativeGetPublicKey(handle: Long): ByteArray?
    private external fun nativeIsHandshakeComplete(handle: Long): Boolean
    private external fun nativeWriteMessage(handle: Long, payload: ByteArray): ByteArray?
    private external fun nativeReadMessage(handle: Long, message: ByteArray): ByteArray?
    private external fun nativeEncrypt(handle: Long, plaintext: ByteArray): ByteArray?
    private external fun nativeDecrypt(handle: Long, ciphertext: ByteArray): ByteArray?
    private external fun nativeGetLastError(handle: Long): Int
}

/**
 * Extension functions for convenience
 */
object NoiseProtocol {
    /**
     * Perform a complete handshake between two sessions
     */
    fun performHandshake(initiator: NoiseSession, responder: NoiseSession) {
        // Message 1: initiator -> responder
        val msg1 = initiator.writeMessage()
        responder.readMessage(msg1)
        
        // Message 2: responder -> initiator
        val msg2 = responder.writeMessage()
        initiator.readMessage(msg2)
        
        // Message 3: initiator -> responder
        val msg3 = initiator.writeMessage()
        responder.readMessage(msg3)
        
        // Verify both are in transport mode
        if (!initiator.isHandshakeComplete || !responder.isHandshakeComplete) {
            throw NoiseException.HandshakeFailed()
        }
    }
}

/**
 * Extension functions for ByteArray
 */
fun ByteArray.toHexString(): String {
    return joinToString("") { "%02x".format(it) }
}

fun String.hexToByteArray(): ByteArray {
    check(length % 2 == 0) { "Hex string must have even length" }
    
    return chunked(2)
        .map { it.toInt(16).toByte() }
        .toByteArray()
}

File: examples/ios/Sources/NoiseSession.swift
import Foundation

/// Noise session mode
public enum NoiseMode: Int32 {
    case initiator = 0
    case responder = 1
}

/// Noise protocol errors
public enum NoiseError: Error {
    case invalidParameter
    case outOfMemory
    case handshakeFailed
    case encryptionFailed
    case decryptionFailed
    case bufferTooSmall(needed: Int)
    case invalidState
    case protocolError
    case unknown(Int32)
    
    init(code: Int32) {
        switch code {
        case 1: self = .invalidParameter
        case 2: self = .outOfMemory
        case 3: self = .handshakeFailed
        case 4: self = .encryptionFailed
        case 5: self = .decryptionFailed
        case 6: self = .bufferTooSmall(needed: 0)
        case 7: self = .invalidState
        case 8: self = .protocolError
        default: self = .unknown(code)
        }
    }
}

/// Swift wrapper for Noise protocol sessions
public class NoiseSession {
    private var session: OpaquePointer?
    
    /// Create a new Noise session
    public init(mode: NoiseMode) throws {
        var error: Int32 = 0
        session = noise_session_new(mode.rawValue, &error)
        
        if error != 0 {
            throw NoiseError(code: error)
        }
        
        if session == nil {
            throw NoiseError.outOfMemory
        }
    }
    
    /// Create a new Noise session with a specific private key
    public init(mode: NoiseMode, privateKey: Data) throws {
        guard privateKey.count == 32 else {
            throw NoiseError.invalidParameter
        }
        
        var error: Int32 = 0
        session = privateKey.withUnsafeBytes { keyBytes in
            noise_session_new_with_key(
                keyBytes.bindMemory(to: UInt8.self).baseAddress,
                privateKey.count,
                mode.rawValue,
                &error
            )
        }
        
        if error != 0 {
            throw NoiseError(code: error)
        }
        
        if session == nil {
            throw NoiseError.outOfMemory
        }
    }
    
    deinit {
        if let session = session {
            noise_session_free(session)
        }
    }
    
    /// Get the local public key
    public var publicKey: Data {
        get throws {
            guard let session = session else {
                throw NoiseError.invalidState
            }
            
            var keyBuffer = Data(count: 32)
            var keyLen = keyBuffer.count
            
            let result = keyBuffer.withUnsafeMutableBytes { buffer in
                noise_get_public_key(
                    session,
                    buffer.bindMemory(to: UInt8.self).baseAddress,
                    &keyLen
                )
            }
            
            if result != 0 {
                throw NoiseError(code: result)
            }
            
            return keyBuffer
        }
    }
    
    /// Check if handshake is complete
    public var isHandshakeComplete: Bool {
        guard let session = session else { return false }
        return noise_is_handshake_complete(session) != 0
    }
    
    /// Write a handshake or transport message
    public func writeMessage(_ payload: Data = Data()) throws -> Data {
        guard let session = session else {
            throw NoiseError.invalidState
        }
        
        // Allocate buffer for output (max message size + overhead)
        var outputBuffer = Data(count: 65535)
        var outputLen = outputBuffer.count
        
        let result = outputBuffer.withUnsafeMutableBytes { output in
            if payload.isEmpty {
                return noise_write_message(
                    session,
                    nil,
                    0,
                    output.bindMemory(to: UInt8.self).baseAddress,
                    &outputLen
                )
            } else {
                return payload.withUnsafeBytes { input in
                    noise_write_message(
                        session,
                        input.bindMemory(to: UInt8.self).baseAddress,
                        payload.count,
                        output.bindMemory(to: UInt8.self).baseAddress,
                        &outputLen
                    )
                }
            }
        }
        
        if result == 6 { // BUFFER_TOO_SMALL
            throw NoiseError.bufferTooSmall(needed: outputLen)
        } else if result != 0 {
            throw NoiseError(code: result)
        }
        
        return outputBuffer.prefix(outputLen)
    }
    
    /// Read a handshake or transport message
    public func readMessage(_ message: Data) throws -> Data {
        guard let session = session else {
            throw NoiseError.invalidState
        }
        
        var outputBuffer = Data(count: 65535)
        var outputLen = outputBuffer.count
        
        let result = outputBuffer.withUnsafeMutableBytes { output in
            message.withUnsafeBytes { input in
                noise_read_message(
                    session,
                    input.bindMemory(to: UInt8.self).baseAddress,
                    message.count,
                    output.bindMemory(to: UInt8.self).baseAddress,
                    &outputLen
                )
            }
        }
        
        if result == 6 { // BUFFER_TOO_SMALL
            throw NoiseError.bufferTooSmall(needed: outputLen)
        } else if result != 0 {
            throw NoiseError(code: result)
        }
        
        return outputBuffer.prefix(outputLen)
    }
    
    /// Encrypt a message (transport mode only)
    public func encrypt(_ plaintext: Data) throws -> Data {
        guard let session = session else {
            throw NoiseError.invalidState
        }
        
        var ciphertext = Data(count: plaintext.count + 16) // Add AEAD tag
        var cipherLen = ciphertext.count
        
        let result = ciphertext.withUnsafeMutableBytes { cipher in
            plaintext.withUnsafeBytes { plain in
                noise_encrypt(
                    session,
                    plain.bindMemory(to: UInt8.self).baseAddress,
                    plaintext.count,
                    cipher.bindMemory(to: UInt8.self).baseAddress,
                    &cipherLen
                )
            }
        }
        
        if result != 0 {
            throw NoiseError(code: result)
        }
        
        return ciphertext.prefix(cipherLen)
    }
    
    /// Decrypt a message (transport mode only)
    public func decrypt(_ ciphertext: Data) throws -> Data {
        guard let session = session else {
            throw NoiseError.invalidState
        }
        
        var plaintext = Data(count: ciphertext.count)
        var plainLen = plaintext.count
        
        let result = plaintext.withUnsafeMutableBytes { plain in
            ciphertext.withUnsafeBytes { cipher in
                noise_decrypt(
                    session,
                    cipher.bindMemory(to: UInt8.self).baseAddress,
                    ciphertext.count,
                    plain.bindMemory(to: UInt8.self).baseAddress,
                    &plainLen
                )
            }
        }
        
        if result != 0 {
            throw NoiseError(code: result)
        }
        
        return plaintext.prefix(plainLen)
    }
}

// MARK: - Convenience Extensions

extension NoiseSession {
    /// Perform a complete handshake as initiator
    public static func performHandshake(
        initiator: NoiseSession,
        responder: NoiseSession
    ) throws {
        // Message 1: initiator -> responder
        let msg1 = try initiator.writeMessage()
        _ = try responder.readMessage(msg1)
        
        // Message 2: responder -> initiator
        let msg2 = try responder.writeMessage()
        _ = try initiator.readMessage(msg2)
        
        // Message 3: initiator -> responder
        let msg3 = try initiator.writeMessage()
        _ = try responder.readMessage(msg3)
        
        // Verify both are in transport mode
        guard initiator.isHandshakeComplete && responder.isHandshakeComplete else {
            throw NoiseError.handshakeFailed
        }
    }
}

// MARK: - Hex String Support

extension Data {
    /// Convert data to hex string
    var hexString: String {
        map { String(format: "%02hhx", $0) }.joined()
    }
    
    /// Create data from hex string
    init?(hexString: String) {
        let len = hexString.count / 2
        var data = Data(capacity: len)
        var index = hexString.startIndex
        
        for _ in 0..<len {
            let nextIndex = hexString.index(index, offsetBy: 2)
            guard let byte = UInt8(hexString[index..<nextIndex], radix: 16) else {
                return nil
            }
            data.append(byte)
            index = nextIndex
        }
        
        self = data
    }
}